File: C/Linked List/Singly Linked List/SinglyLL009.c
/*
=====================================================================

Program Name    : Singly Linked List - Using Pointer Alias with typedef
Program Number  : 009
Description     : This program demonstrates the use of typedef to
                  create aliases for both the node structure and
                  node pointer. It simplifies pointer declarations
                  and improves code readability in linked list
                  implementations.

Concepts Used   : Structure, Self-Referential Structure,
                  typedef, Pointer Alias,
                  Head Pointer, Node Linking

Input           : None

Output          : Displays the data stored in all linked nodes.

Date            : July 2026
Language        : C
Author          : Atharv Tushar Bhosale

=====================================================================
*/

#include<stdio.h>

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

int main()
{
    PNODE head = NULL;
    NODE obj1, obj2, obj3;

    head = &obj1;

    obj1.data = 11;
    obj1.next = &obj2;

    obj2.data = 21;
    obj2.next = &obj3;

    obj3.data = 51;
    obj3.next = NULL;

    printf("%d\n", head->data);
    printf("%d\n", head->next->data);
    printf("%d\n", head->next->next->data);

    return 0;
}

/*
=====================================================================

Expected Output :
11
21
51

Explanation :
Two typedef aliases are created:

NODE  -> Represents 'struct node'
PNODE -> Represents 'struct node *'

This allows node pointers to be declared using
PNODE instead of repeatedly writing
'struct node *'.

Without typedef :
struct node *head;

With typedef :
PNODE head;

Operations Performed :
1. Created a structure alias (NODE).
2. Created a pointer alias (PNODE).
3. Declared a head pointer using PNODE.
4. Linked three nodes together.
5. Traversed the linked list through head.
6. Displayed data from all nodes.

Conclusion :
Using typedef for pointer declarations improves
code readability, reduces complexity, and makes
linked list implementations easier to understand
and maintain.

=====================================================================
*/

File: Java/Linked List/Singly Linked List/SinglyLL009.java
/*
=====================================================================

Name            : Singly Linked List - InsertFirst and InsertLast
Program Number  : 009

Description     : This program demonstrates the implementation of
                  InsertFirst() and InsertLast() operations in a
                  Singly Linked List. Nodes can be inserted at the
                  beginning and at the end of the linked list.

Concepts Used   : Class, Object Creation, Constructor,
                  Self-Referential Class,
                  Dynamic Memory Allocation,
                  InsertFirst Operation,
                  InsertLast Operation,
                  Linked List Traversal

Input           : 11, 21, 51, 101, 111, 121

Output          : Nodes inserted successfully in the
                  Singly Linked List

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale

=====================================================================
*/

class node
{
    public int data;
    public node next;

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL()
    {
        this.first = null;
        this.iCount = 0;
    }

    public void Display()
    {}

    public int Count()
    {
        return iCount;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);
        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }
    }

    public void InsertLast(int No)
    {
        node newn = new node(No);
        node temp = null;

        newn.data = No;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }
    }

    public void InsertAtPos(int iNo,int iPos)
    {}

    public void DeleteFirst()
    {}

    public void DeleteLast()
    {}

    public void DeleteAtPos(int iPos)
    {}
}

class SinglyLL009
{
    public static void main(String A[])
    {
        SinglyLL sobj = new SinglyLL();

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);
    }
}

/*
=====================================================================

Summary         : Implemented insertion operations at the
                  beginning and end of the linked list.

Linked List     : 11 -> 21 -> 51 -> 101 -> 111 -> 121 -> NULL

Learning        : InsertFirst requires only head manipulation,
                  whereas InsertLast requires traversal up to
                  the last node before attaching the new node.

=====================================================================
*/
